docs/lazy-adr/adr-005-da-modules-and-components.md

```diff
@@ -0,0 +1,122 @@
+is in its name - pull. That is, pull DA pulls Data by requesting it from the "network" mainly using
+[kDHT from IPFS/libp2p](https://github.com/libp2p/go-libp2p-kad-dht).
+
+LL's specific DHT usage is described in high-level [here](https://github.com/lazyledger/lazyledger-core/issues/395).
+
+All pull related types and logic is going to be implemented in it's own package `dah/pull`
+
+#### DAProvider
+For DHT related operation a `DAProvider` is introduced with further semantics:
+```go
+type DAProvider interface {
+    // Provide announces to the network that the node calling it hosts data addressed by given DAHeader.
+    // Announcement lives for 24h by default.
+    Provide(context.Context, DAHeader) error
+    // Reprovide periodically(every 24h by default) executes Provide over given DAHeader.
+    // It persists reporvide requirement in datastore to continue reproviding after app restarts.
+    Reprovide(DAHeader) error
+    // Unprovide prevents further reprovides for a given DAHeader
+    Unprovide(DAHeader) error
+}
+```
+
+// TODO: Add to specs as well
+We also define several system-wide parameters related to Provider and DHT:
+* `provideValidity=24h` - lifespan of a providing record.
+* `kBucketSize=20` - configures bucket size of the kademlia routing table.
+* `resiliency=3` - the number of peers closest to a target that must have responded in order for a given DHT query to
+  complete.
+* `alpha=32` - configures the number of concurrent requests for a given DHT query.
+* `protocolPrefix="/celestia"` - protocol prefix to distinguish our DHT network
+
+#### PullDA
+The DA implementation itself.
+
+##### Put
+As from the method description of the interface, pull version of `Put` ensures `Data` is available on the network.
+In terms of "pulling" that means requestable and retreivable by any peer from the node calling the `Put`.
+
+Logic simply exucutes `Provider.Provide` and `Provider.Reprovide` for every DA Header column and row roots. This ensures
+capabilities described above initially and over time.
+
+##### Get
+
+### Push DA Components
+Push DA is a new DA implementation meant to solve slow data retreivability problem by proactively broadcasting content
+rather than requesting it as in pull approach. For broadcasting data to network the plan is to rely on
+[libp2p PubSub protocol](https://github.com/libp2p/go-libp2p-pubsub) which provides simple interface and strong
+guarantees for content delivery, while managing peer exchange and network topoly formation.
+
+PubSub provides a notion of topics what allows peers to exchange data contextualized with a specific string topic. Push
+DA will allow setting custom topic for users. For LL we will use the `/celestia` topic.
+
+#### PushDA
+
+##### Put
+`Put` in push schema takes all the shares of an extended square, encodes every share along with a `(col;row)` coordinate
+together with `DataHash` and publishes them to the `/celestia` topic.
```

Comment:
I hope you don't mind the question: doesn't this introduce a DoS attack vector? because you can only validate if shares match a header on a row or column basis (if you have the DAHeader at hand, otherwise you need enough shares to reconstruct the whole block until you can be sure). @adlerjohn raised the concern that with presumably really large blocks this could cause problems.

Reply:
> doesn't this introduce a DoS attack vector?

@liamsi, Note that there is no difference between spamming shares and rows/cols, as you can generate dummy data and spam the network for both with any amount. The only way to mitigate that is to use [peer scoring](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#peer-scoring). I'll describe how we can use this soon.

Reply:
> Note that there is no difference between spamming shares and rows/cols, as you can generate dummy data and spam the network for both with any amount

Yeah, but it makes a difference how much data you must read before you can actually go and score a peer accordingly. In the case of rows you could after one garbage row, in the case of shares (and only having the data root) you'd need to see all shares first.

Reply:
@liamsi, It is true in case we use DAHeader, but later on we want to have DataHash only, and in such a case you will get the same situation as with sample granularity. DAHeader is essentially an inner node, so the question is - do we need to broadcast inner nodes?

Reply:
> DAHeader is essentially an inner node, so the question is - do we need to broadcast inner nodes?

While that statement is true, I still see the DAHeader as a special case of inner nodes. We might want to broadcast the DAHeader but not all inner nodes. I need to think about this.

Reply:
That's an open question for me and from first sight, if inner nodes are broadcasted, nodes would need to load a bunch of data, in the same way, to verify it, as PubSub does not guarantee delivery order. Though we can assume that broadcasting happens from root to leaves, so less data needs to be loaded to do verification, but the problem you've mentioned will stay anyway.

Reply:
However, I think that's not a huge problem that we can't do immediate verification for a sample/row/inner node. Any validator can easily allocate maxUnextendedBlockSize-shareSize RAM even for garbage that later is freed with an objective penalty for a spamming peer.

Reply:
So I would still stick to transmitting all the shares with recovery as early as possible and without inner node broadcasting, as bandwidth optimization.

Reply:
Yeah sure, for block data gossiping, we should _not_ broadcast all inner nodes. Only if we want to e.g. check earlier without having to download all data, then we could also provide the DAHeader (which only would make sense if the granularity is rows, and not shares–not sure what the best granularity is; just thinking about potential attacks that could be specific to the very-large-blocks-case).

Also, one question: when you say shares, do you mean all shares, or, only the original shares?

Reply:
Chipping in here to note requiring 1/4 of the EDS (i.e. an ODS worth of shares) to be downloaded without verification will be a DoS vector. Not a RAM resource exhaustion, but a network bandwidth exhaustion.

@Wondertan when you say "and publishes them," are shares published individually, or an entire block? From what you wrote in this comment thread, it looks like individual shares, for which Merkle proofs are downloaded? Not sure.

Reply:
> Chipping in here to note requiring 1/4 of the EDS (i.e. an ODS worth of shares) to be downloaded without verification will be a DoS vector. Not a RAM resource exhaustion, but a network bandwidth exhaustion.

It is until you do verification, penalize spamming peer, stopping further communications with him, thus preventing DoS.

Reply:
> when you say "and publishes them," are shares published individually, or an entire block?

@adlerjohn, individually

Reply:
If we can't verify the block until we fully received or recovered it, then there is no difference between sending it fully or by shares or by other pieces in terms of verification. However, sending shares in a batch could save more bandwidth, as we can append DataHash to batch instead of per share.

Reply:
The batch size can be maxed out to max PubSub message size. [The default is 1mb](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L27) and I am not aware of any rationale behind that, though there should be the most optimal message size for our use case.